File: src/data_analysis/translate_wangchun_r/guo_main.py
"""
@author  : MG
@Time    : 2020/8/30 9:05
@File    : guo_main.py
@contact : mmmaaaggg@163.com
@desc    : 用于
"""
import sqlite3
import pandas as pd
import numpy as np


class PortConstructor:

    def __init__(self,
                 nsel, nmemory, nalpha, nstyle, risk_tolerance,
                 sector_deviation=0.025, cvar_alpha=0.05, risk_measure="CVaR", cov_estimator="cov"
                 ):
        """

        :param nsel: we select nsel number of stocks
        :param nmemory: window size for Sigma estimation
        :param nalpha: window size for expected return estimation
        :param nstyle: number of style constraints
        :param risk_tolerance: risk tolerance
        :param sector_deviation:
        :param cvar_alpha:
        :param risk_measure: "Risk_Adjusted_Return"/"Risk_Parity"/"Minimum_Tail_Dependent"/"Minimum_Variance"/"CVaR"
        :param cov_estimator: "cov"/"lpmEstimator"/"slpmEstimator"/"kendallEstimator"/"spearmanEstimator"/"shrinkEstimator"
        """
        self.nsel = nsel
        self.nmemory = nmemory
        self.nalpha = nalpha
        self.nstyle = nstyle
        self.risk_tolerance = risk_tolerance
        self.sector_deviation = sector_deviation
        self.cvar_alpha = cvar_alpha
        self.risk_measure = risk_measure
        self.cov_estimator = cov_estimator
        self.wup = 5 / nsel  # upper bound for each stock #
        self.wdown = 1 / nsel / 3  # lower bound for each stock

        self.md = None
        self.fv_df = None
        self.beta_df = None

        self.date_s = None
        self.sector_class2_df = None
        self.sector_unique_list = None

        # 从 DB_FuJuDailybar.db 获取的数据
        self.code_list = None
        self.volume_df = None
        self.cc_df = None
        self.log_cc_df = None

        self.portfolio_weights = None

    def get_md(self, file_path):
        conn = sqlite3.connect(file_path)

        def tables_in_sqlite_db(conn):
            cursor = conn.execute("SELECT name FROM sqlite_master WHERE type='table';")
            tables = [
                v[0] for v in cursor.fetchall()
                if v[0] != "sqlite_sequence"
            ]
            cursor.close()
            return tables

        tables = tables_in_sqlite_db(conn)
        n = len(tables)
        self.code_list = []
        df = pd.read_sql_query("SELECT * FROM X000001", conn)
        T = df.shape[0]
        Open = np.zeros(shape=(T, n))
        High = np.zeros(shape=(T, n))
        Low = np.zeros(shape=(T, n))
        Close = np.zeros(shape=(T, n))
        LiquidSize = np.zeros(shape=(T, n))
        self.volume_df = np.zeros(shape=(T, n))
        Size = np.zeros(shape=(T, n))
        for i in range(0, (n - 1)):
            df = pd.DataFrame(pd.read_sql_query("SELECT * FROM " + tables[i], conn))
            self.code_list.append(tables[i][1:])
            Open[:, i] = df["Open"]  # A.Open
            High[:, i] = df["High"]
            Low[:, i] = df["Low"]
            Close[:, i] = df["Close"]
            LiquidSize[:, i] = df["LiquidSize"]
            Size[:, i] = df["Size"]
            self.volume_df[:, i] = df["Volume"]

        # CC = np.zeros(shape=(T, n))
        # for i in range(0, (n - 1)):
        #     CC[2:T, i] = Close[2:T, i] / Close[1:(T - 1), i] - 1

        close_df = pd.DataFrame(Close, columns=self.code_list)
        self.cc_df = (close_df.pct_change().iloc[1:, :] / close_df.iloc[:-1, :]).fillna(0)
        log_close_df = np.log(close_df)
        self.log_cc_df = log_close_df.pct_change().iloc[1:, :].fillna(0)

    def load_fv(self, file_path):
        """
        加载 factor value
        :param file_path:
        :return:
        """
        df = pd.read_csv(file_path, index_col=[0])
        self.fv_df = df.fillna(-100)

    def load_beta(self, file_path):
        """
        加载 beta
        :param file_path:
        :return:
        """
        self.beta_df = pd.read_csv(file_path, index_col=[0])

    def load_sector_class_df(self, file_path):
        sector_class_dic = {}
        with sqlite3.connect(file_path) as conn:
            for num, code in enumerate(self.code_list, start=1):
                table_name = f"X{code[:6]}"
                # print(f"{num:4d}) {code} -> {table_name}")
                stock_df = pd.read_sql(f"select * from {table_name}", conn)
                if self.date_s is None:
                    self.date_s = pd.to_datetime(stock_df['Date'])
                sector_class_dic[code] = stock_df["TYPE_ID"]
                # if num > 3:  # 调试使用
                #     break

        sector_class_df = pd.DataFrame(sector_class_dic)
        sector_class_df.fillna(0)
        #          000001.SZ     000002.SZ     000004.SZ     000005.SZ
        # 0     1.031721e+10  1.031723e+10  1.031718e+10  1.031723e+10
        # 1     1.031721e+10  1.031723e+10  1.031718e+10  1.031723e+10
        # 2     1.031721e+10  1.031723e+10  1.031718e+10  1.031723e+10
        # 3     1.031721e+10  1.031723e+10  1.031718e+10  1.031723e+10
        # 4     1.031721e+10  1.031723e+10  1.031718e+10  1.031723e+10
        #             ...           ...           ...           ...
        # 3309  1.031721e+10  1.031723e+10  1.031718e+10  1.031704e+10
        # 3310  1.031721e+10  1.031723e+10  1.031718e+10  1.031704e+10
        # 3311  1.031721e+10  1.031723e+10  1.031718e+10  1.031704e+10
        # 3312  1.031721e+10  1.031723e+10  1.031718e+10  1.031704e+10
        # 3313  1.031721e+10  1.031723e+10  1.031718e+10  1.031704e+10
        # 取前8位
        self.sector_class2_df = sector_class_df.applymap(lambda x: str(int(x))[:8])  # SectorClass2
        self.sector_unique_list = [_ for _ in set(self.sector_class2_df.to_numpy().flatten()) if _ != '0']
        self.sector_unique_list.sort()  # SectorU
        # sector_count = len(sector_unique_list)  # nsector

    def construct(self):
        self.portfolio_weights = np.zeros((self.fv_df.shape[0] + 1, self.fv_df.shape[1]))
        nstart = self.nmemory * 2 + 1
        # H is covariance matrix; Alpha is expected return
        for t in range(nstart, self.fv_df.shape[0]):
            fv_s = self.fv_df.iloc[t, :]
            matched_nsel = fv_s.rank(ascending=False) < self.nsel
            matched_nsel_indexes = np.where(matched_nsel)[0]  # jy
            if self.risk_measure == "Minimum_Tail_Dependent":
                H = np.zeros(1)
                if self.cov_estimator == "cov":
                    pass
                elif self.cov_estimator == "lpmEstimator":
                    pass
                elif self.cov_estimator == "slpmEstimator":
                    pass
                elif self.cov_estimator == "kendallEstimator":
                    pass
                elif self.cov_estimator == "spearmanEstimator":
                    pass
                elif self.cov_estimator == "shrinkEstimator":
                    pass
                else:
                    raise KeyError(f'cov_estimator = {self.cov_estimator} 不合法')

                if np.any(np.isnan(H)):
                    H = self.cc_df.iloc[(t - self.nmemory + 1):(t + 1), matched_nsel_indexes].cov()

                h_diag = np.diag(H)
                if np.any(h_diag == 0):  # min(diag(H))==0
                    # diag(H)[which(diag(H)==0)]=mean(diag(H)[which(diag(H)!=0)])
                    for n in np.where(h_diag == 0)[0]:
                        H[n, n] = np.mean(h_diag[h_diag != 0])

                SD = h_diag ** 0.5

            elif self.risk_measure == "Minimum_Tail_Dependent":
                pass
            else:
                alpha = np.zeros(self.nsel)
                for num, sector in enumerate(self.sector_unique_list):
                    matched_sector = (self.sector_class2_df.iloc[-1, :] == sector) & matched_nsel
                    matched_sector_indexes = np.where(matched_sector)[0]  # jz
                    matched_sector_of_nsel_indexes = np.where(np.isin(matched_nsel_indexes, matched_sector_indexes))
                    if np.sum(matched_sector) > 0:
                        alpha[matched_sector_of_nsel_indexes] = np.sum(
                            self.cc_df.iloc[(t - self.nalpha + 1):(t + 1), matched_sector_indexes]
                        ) / np.sum(
                            self.volume_df[(t - self.nalpha + 1):(t + 1), matched_sector_indexes] > 0
                        )

                    if np.isnan(alpha[matched_sector_of_nsel_indexes]):
                        raise ValueError(
                            f"{num}) {sector} 在 {np.where(np.isnan(alpha[matched_sector_of_nsel_indexes]))[0]} 出现 nan")

                    op = self.Constrained_OneStep_OptimFunc5()
                    self.portfolio_weights[t + 1, matched_nsel_indexes] = op['w_opt']  # Wopt[t+1,jy]=op$w_opt;

    def do_job(self):
        pass

File: src/data_analysis/translate_wangchun_r/GuoMain.R
   ### ###
   library("data.table");
   path=PasteAll(c(Disk,":/Rwork/AlphaStrategy/Graph/StrategyResult/EquityCurve_Beta.csv"),"");
   Beta=data.table::fread(path,header=TRUE,fill=TRUE);
   path=PasteAll(c(Disk,":/Rwork/AlphaStrategy/Data/DataHis/equal_weighted.csv"),"");  
   Data=as.data.frame(read.csv(path,header=TRUE,sep=",",encoding="UTF-8"));
   names(Data)[1]="Date";
   # Factor Values #
   FV=array(0,c(T,n));
   for (i in 1:n){
        FV[1:T,i]=Data[,i+1];
   }
   FV[which(is.na(FV))]=-100;
   CC=array(0,c(T,n));        # Close-to-Close return #
   LogCC=array(0,c(T,n));     # Log return #
   for (i in 1:n){
        CC[2:T,i]=Close[2:T,i]/Close[1:(T-1),i]-1;
        LogCC[2:T,i]=log(Close[2:T,i])-log(Close[1:(T-1),i]);
   }                          
   CC[which(is.na(CC))]=0;
   Wopt=array(0,c(T+1,n));    # optimal portfolio weight, T=number of trading days, n=number of stocks #
   nsel=300;                  # we select nsel number of stocks 
   nmemory=90;                # window size for Sigma estimation 
   nalpha=10;                 # window size for expected return estimation 
   nstyle=2;                  # number of style constraints #
   wup=5/nsel;                # upper bound for each stock #
   wdown=1/nsel/3;            # lower bound for each stock
   lambda=1;                  # risk tolerance 
   sectorDeviation=0.025;     # 
   CVaR_alpha=0.05;
   nstart=nmemory*2+1;
   RiskMeasure="CVaR";       # "Risk_Adjusted_Return"/"Risk_Parity"/"Minimum_Tail_Dependent"/"Minimum_Variance"/"CVaR";
   CovEstimator="cov";                   # "cov"/"lpmEstimator"/"slpmEstimator"/"kendallEstimator"/"spearmanEstimator"/"shrinkEstimator";
   source("NeweyWestFunc.R");
   Beta=rbind(array(0,c(2*nmemory+1,n)),as.matrix(Beta[,2:(n+1)]));  # rbind is row combine, cbind is column combine; rbind(array(1,c(2,2)),array(2,c(3,2)))
   Beta[nstart,]=Beta[nstart+1,];
   source("Constrained_OneStep_OptimFunc5.R");
   # H is covariance matrix; Alpha is expected return 
   for (t in nstart:T){
        cat("t",t,"\n");
        jy=which(rank(-FV[t,])<=nsel);   # factor value 排序取最大值 
        if (RiskMeasure!="Minimum_Tail_Dependent"){
            if (any(is.na(H)))   H=cov(CC[(t-nmemory+1):t,jy]);
            if (min(diag(H))==0) diag(H)[which(diag(H)==0)]=mean(diag(H)[which(diag(H)!=0)]);
            SD=diag(H)^0.5;
        } 
        if (RiskMeasure=="Minimum_Tail_Dependent"){
            H=FRAPO::tdc(CC[(t-nmemory+1):t,jy],method="EmpTC",lower=TRUE);
            SD=diag(cov(CC[(t-nmemory+1):t,jy]))^0.5;
            if (any(SD==0)) SD[which(SD==0)]=mean(SD[which(SD!=0)]);
        }
        Alpha=matrix(0,nsel,1);
        if (is.element(RiskMeasure,c("Minimum_Tail_Dependent","Minimum_Variance"))==FALSE){
            for (i in 1:nsector){
                 jz=intersect(jy,which(is.element(SectorClass2[t-1,],SectorU[i])));
                 lz=length(jz);
                 js=which(is.element(jy,jz));
                 if (lz>1) Alpha[js]=sum(CC[(t-nalpha+1):t,jz])/length(which(Volume[(t-nalpha+1):t,jz]>0));
                 if (any(is.na(Alpha[js]))) stop("Alpha[js]");
            } 
        }
        op=Constrained_OneStep_OptimFunc5(RiskMeasure,t,jy,CC,IndexRe,Cons,Code,nmemory,nalpha,nstyle,H,SD,nsector,Alpha,SectorClass2,SectorU,IndexSector500,Size,nsel,w0,wup,wdown,lambda,sectorDeviation,Beta,CVaR_alpha);
        Wopt[t+1,jy]=op$w_opt;
   }






   
   PortRe=array(0,T);
   PortRe2=array(0,T);
   PortRe[1:nstart]=IndexRe[1:nstart];
   PortRe[(nstart+1):T]=rowSums(Wopt[(nstart+1):T,]*CC[(nstart+1):T,]);
   PortRe2[(nstart+1):T]=PortRe[(nstart+1):T]-IndexRe[(nstart+1):T];
   PortV=cumprod(1+PortRe[(nstart+1):T]);
   PortV2=cumprod(1+PortRe2[(nstart+1):T]);
   plot(x=as.Date(Date)[(nstart+1):T],y=PortV,type="l",col="red",xlab="Trading Day",ylab="Value");
   grid();
   plot(x=as.Date(Date)[(nstart+1):T],y=PortV2,type="l",col="red",xlab="Trading Day",ylab="Value");
   grid();
   plot(x=as.Date(Date)[(nstart+1):T],y=log(PortV),type="l",col="red",xlab="Trading Day",ylab="Value");
   grid();
   plot(x=as.Date(Date)[(nstart+1):T],y=log(PortV2),type="l",col="red",xlab="Trading Day",ylab="Value");
   grid();
